### PR TITLE
Wallet version and network version detection

### PR DESCRIPTION
**Pull request by JustynaBroniszewska**

### CODE REVIEW

File: universal-login-commons/lib/integration/ethereum/fetchHardforkVersion.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    case 'unknown':
+      return 'constantinople';
+    case 'homestead':
+      const blockNumber = await provider.getBlockNumber();
+      if (blockNumber < ISTANBUL_BLOCK_NUMBER) {
+        return 'constantinople';
+      }
+      return 'istanbul';
+  }
+};
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/integration/ethereum/fetchHardforkVersion.ts]: 👍

File: universal-login-commons/test/integration/ethereum/fetchHardforkVersion.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import {providers} from 'ethers';
+import {fetchHardforkVersion, ISTANBUL_BLOCK_NUMBER} from '../../../lib/integration/ethereum/fetchHardforkVersion';
+import {mockProviderWithBlockNumber} from '../../helpers/mockProvider';
+
+describe('fetchHardforkVersion', async () => {
+  const itReturnsHardforkVersion = (network: string, hardfork: string, blockNumber?: number) =>
+    it(`returns ${hardfork} for ${network} ${blockNumber && `for block number equals ${blockNumber}`}`, async () => {
+      const provider = mockProviderWithBlockNumber(network, blockNumber || 0);
+      expect(await fetchHardforkVersion(provider as providers.Provider)).to.eq(hardfork);
+    });
```
Comment (marekkirejczyk) [inline universal-login-commons/test/integration/ethereum/fetchHardforkVersion.ts]: I feel like we are abusing this pattern :)

File: universal-login-sdk/lib/api/DeployedWallet.ts [hunk 3]
```diff
@@ -106,7 +106,7 @@ export class DeployedWallet implements ApplicationWallet {
       from: this.contractAddress,
       data,
     };
-    return this.sdk.execute(message, this.privateKey);
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/api/DeployedWallet.ts]: 👍

File: universal-login-sdk/lib/api/sdk.ts [hunk 4]
```diff
@@ -182,7 +183,7 @@ class UniversalLoginSDK {
     );
   }
 
-  async execute(message: Partial<Message>, privateKey: string): Promise<Execution> {
+  async execute(message: PartialRequired<Message, 'from'>, privateKey: string): Promise<Execution> {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/api/sdk.ts]: 👍

**Comment by marekkirejczyk** [comment]
So far, so good.

File: universal-login-commons/lib/core/constants/constants.ts [hunk 5]
```diff
@@ -9,3 +9,7 @@ export const ETHER_NATIVE_TOKEN = {
 export const EMPTY_DATA = utils.formatBytes32String('0');
 
 export const ONE_SIGNATURE_GAS_COST = utils.bigNumberify('4420');
+
+export const ACTUAL_WALLET_VERSION = 'beta2';
```
Comment (xitronix) [inline universal-login-commons/lib/core/constants/constants.ts]: CURRENT

File: universal-login-commons/lib/integration/ethereum/fetchHardforkVersion.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    case 'unknown':
+      return 'constantinople';
+    case 'homestead':
+      const blockNumber = await provider.getBlockNumber();
+      if (blockNumber < ISTANBUL_BLOCK_NUMBER) {
+        return 'constantinople';
+      }
+      return 'istanbul';
+    default:
+      return 'constantinople';
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/integration/ethereum/fetchHardforkVersion.ts]: throw excption

File: universal-login-contracts/lib/estimateGas.ts [hunk 7]
```diff
@@ -14,13 +14,13 @@ export const calculateFinalGasLimit = (gasCall: utils.BigNumberish, gasBase: uti
 export const calculatePaymentOptions = (msg: SignedMessagePaymentOptions) =>
   ({gasLimit: calculateFinalGasLimit(msg.gasCall, msg.gasBase)});
 
-export const calculateGasBase = (message: Omit<Message, 'gasLimit'>) => {
+export const calculateGasBase = (message: Omit<Message, 'gasLimit'>, networkVersion: ChainVersion, walletVersion: WalletVersion) => {
```
Comment (marekkirejczyk) [inline universal-login-contracts/lib/estimateGas.ts]: To consider next PR: join `networkVersion` and `walletVersion` into `gasCalculationEnvironment`?

File: universal-login-commons/lib/core/utils/messages/computeGasData.ts [hunk 8]
```diff
@@ -6,7 +6,7 @@ export const NON_ZERO_BYTE_GAS_COST = 68;
 export const NEW_NON_ZERO_BYTE_GAS_COST = 16;
 export const GAS_FIXED = '50000';
 
-type ChainVersion = 'istanbul' | 'constantinople';
+export type ChainVersion = 'istanbul' | 'constantinople';
 
 export class GasComputation {
```
Comment (xitronix) [inline universal-login-commons/lib/core/utils/messages/computeGasData.ts]: Move to core/services

File: universal-login-commons/lib/core/utils/messages/computeGasData.ts [hunk 9]
```diff
@@ -6,7 +6,7 @@ export const NON_ZERO_BYTE_GAS_COST = 68;
 export const NEW_NON_ZERO_BYTE_GAS_COST = 16;
 export const GAS_FIXED = '50000';
```
Comment (xitronix) [inline universal-login-commons/lib/core/utils/messages/computeGasData.ts]: Move to constants?

File: universal-login-contracts/lib/estimateGas.ts (hunk 7 above)
Comment (xitronix) [inline universal-login-contracts/lib/estimateGas.ts]: Pick one name chainVersion/NetworkVersion

File: universal-login-commons/lib/core/utils/messages/computeGasData.ts (hunk 8 above)
Comment (JustynaBroniszewska) [reply universal-login-commons/lib/core/utils/messages/computeGasData.ts]: Will do in the following PR

File: universal-login-commons/lib/core/utils/messages/computeGasData.ts (hunk 9 above)
Comment (JustynaBroniszewska) [reply universal-login-commons/lib/core/utils/messages/computeGasData.ts]: Will do in the following PR

File: universal-login-contracts/lib/estimateGas.ts (hunk 7 above)
Comment (JustynaBroniszewska) [reply universal-login-contracts/lib/estimateGas.ts]: 👍

**Pull request merged**